cmd/snap-fde-keymgr/main.go

```diff
@@ -0,0 +1,123 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/secboot"
+	"github.com/snapcore/snapd/secboot/keymgr"
+)
+
+var osStdin io.Reader = os.Stdin
+
+type commonDeviceMixin struct {
+	Device string `long:"device" description:"encrypted device" required:"yes"`
+}
+
+type cmdAddRecoveryKey struct {
+	commonDeviceMixin
+	KeyFile string `long:"key-file" description:"path to recovery key file" required:"yes"`
+}
+
+type cmdRemoveRecoveryKey struct {
+	commonDeviceMixin
+	KeyFile string `long:"key-file" description:"path to recovery key file" required:"yes"`
+}
+
+type cmdChangeEncryptionKey struct {
+	commonDeviceMixin
+}
+
+type options struct {
+	CmdAddRecoveryKey      cmdAddRecoveryKey      `command:"add-recovery-key"`
+	CmdRemoveRecoveryKey   cmdRemoveRecoveryKey   `command:"remove-recovery-key"`
+	CmdChangeEncryptionKey cmdChangeEncryptionKey `command:"change-encryption-key"`
+}
+
+var (
+	keymgrAddRecoveryKeyToLUKSDevice      = keymgr.AddRecoveryKeyToLUKSDevice
+	keymgrRemoveRecoveryKeyFromLUKSDevice = keymgr.RemoveRecoveryKeyFromLUKSDevice
+	keymgrChangeLUKSDeviceEncryptionKey   = keymgr.ChangeLUKSDeviceEncryptionKey
+)
+
+func (c *cmdAddRecoveryKey) Execute(args []string) error {
+	recoveryKey, err := secboot.NewRecoveryKey()
+	if err != nil {
+		return fmt.Errorf("cannot create recovery key: %v", err)
+	}
+	if err := keymgrAddRecoveryKeyToLUKSDevice(c.Device, recoveryKey); err != nil {
+		return fmt.Errorf("cannot add recovery key to LUKS device: %v", err)
+	}
+	if err := ioutil.WriteFile(c.KeyFile, recoveryKey[:], 0600); err != nil {
+		return fmt.Errorf("cannot write recovery key to file: %v", err)
+	}
```

Comment:
as discussed we should try to make this idempotent by writing to file first and then first checking if the file exists and then readding but being happy if the adding says we are adding the same thing, otherwise as I mentioned another strategy is to somehow count the active slots to decide what to do